h2o-automl/src/main/java/ai/h2o/automl/AutoML.java

```diff
@@ -164,76 +166,142 @@ public AutoML() {
 
   public AutoML(Key<AutoML> key, Date startTime, AutoMLBuildSpec buildSpec) {
     super(key == null ? Key.make(buildSpec.project()) : key);
-    _startTime = startTime;
-    _buildSpec = buildSpec;
-    _runCountdown = Countdown.fromSeconds(buildSpec.build_control.stopping_criteria.max_runtime_secs());
-    _modelingPlan = buildSpec.build_models.modeling_plan == null ? defaultModelingPlan : buildSpec.build_models.modeling_plan;
-    _modelingStepsRegistry = new ModelingStepsRegistry();
-
     try {
+      _startTime = startTime;
       _eventLog = EventLog.getOrMake(_key);
-      eventLog().info(Stage.Workflow, "Project: " + projectName());
-
-      handleCVParameters(buildSpec);
-
-      handleReproducibilityParameters(buildSpec);
-
-      handleDatafileParameters(buildSpec);
+      eventLog().info(Stage.Workflow, "Project: "+buildSpec.project());
 
-      handleEarlyStoppingParameters(buildSpec);
-
-      initLeaderboard(buildSpec);
+      validateBuildSpec(buildSpec);
+      _buildSpec = buildSpec;
+      // now that buildSpec is validated, we can assign it: all future logic can now safely access parameters through _buildSpec.
+      _runCountdown = Countdown.fromSeconds(_buildSpec.build_control.stopping_criteria.max_runtime_secs());
 
+      prepareData();
+      initLeaderboard();
       planWork();
-
       _modelingStepsExecutor = new ModelingStepsExecutor(_leaderboard, _eventLog, _runCountdown);
     } catch (Exception e) {
       delete(); //cleanup potentially leaked keys
       throw e;
     }
   }
 
-  private void initLeaderboard(AutoMLBuildSpec buildSpec) {
-    String sort_metric = buildSpec.input_spec.sort_metric;
-    sort_metric = sort_metric == null || StoppingMetric.AUTO.name().equalsIgnoreCase(sort_metric) ? null : sort_metric.toLowerCase();
-    _leaderboard = Leaderboard.getOrMake(_key.toString(), _eventLog, _leaderboardFrame, sort_metric);
-  }
+  private void validateBuildSpec(AutoMLBuildSpec buildSpec) {
+    // Validate all buildSpec parameters here and provide reasonable defaults or parameter cleaning if necessary.
+    // When assigning default or modifying the original value, use `eventLog` to inform the user about the changes.
+    // Please no additional logic!
 
-  private void handleReproducibilityParameters(AutoMLBuildSpec buildSpec) {
-    eventLog().info(Stage.Workflow, "Build control seed: " + buildSpec.build_control.stopping_criteria.seed() +
-            (buildSpec.build_control.stopping_criteria.seed() == -1 ? " (random)" : ""));
+    // Can possibly use this validation entry point to provide parameters validation for Flow later.
+    // Can also reuse ModelBuilder.ValidationMessage for this.
+
+    validateInput(buildSpec.input_spec);
```

Comment:
@seb-h2o when reading over the file I found it slightly confusing that we have `validateInput` and then `prepareData`. So we first validate and then prepare/change data which looks unusual to me.

Reply:
yeah, validation is only about checking that all buildSpec parameters are correct, consistent with each other... its only allowed side effect is to modify some of those params to set some defaults dynamically.
Before validation is completed, you don't know yet if you'll start the training.

As a comparison, you start preparing data when the training part is engaged.

The distinction is clear from a user perspective, like ModelBuilder does. We could make a request to AutoML just to check the parameters, and display messages, highlight wrong parameter fields... this should be fast, and you don't want to start preparing data in this request.

Reply:
@seb-h2o I understand. It is only the fact that within validation of buildSpec we actually do validate Input ( which is data as well) that confused me